source/extensions/filters/network/dubbo_proxy/dubbo_protocol_impl.cc

```diff
@@ -0,0 +1,114 @@
+#include "extensions/filters/network/dubbo_proxy/dubbo_protocol_impl.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace DubboProxy {
+namespace {
+
+constexpr uint16_t MagicNumber = 0xdabb;
+constexpr uint8_t MessageTypeMask = 0x80;
+constexpr uint8_t EventMask = 0x20;
+constexpr uint8_t TwoWayMask = 0x40;
+constexpr uint8_t SerializationTypeMask = 0x1f;
+
+} // namespace
+
+// Consistent with the SerializationType
+bool isValidSerializationType(SerializationType type) {
+  switch (type) {
+  case SerializationType::Hessian:
+  case SerializationType::Json:
+    break;
+  default:
+    return false;
+  }
+  return true;
+}
+
+// Consistent with the ResponseStatus
+bool isValidResponseStatus(ResponseStatus status) {
+  switch (status) {
+  case ResponseStatus::Ok:
+  case ResponseStatus::ClientTimeout:
+  case ResponseStatus::ServerTimeout:
+  case ResponseStatus::BadRequest:
+  case ResponseStatus::BadResponse:
+  case ResponseStatus::ServiceNotFound:
+  case ResponseStatus::ServiceError:
+  case ResponseStatus::ClientError:
+  case ResponseStatus::ServerThreadpoolExhaustedError:
+    break;
+  default:
+    return false;
+  }
+  return true;
+}
+
+void RequestMessageImpl::fromBuffer(Buffer::Instance& data) {
+  ASSERT(data.length() >= DubboProtocolImpl::MessageSize);
+  uint8_t flag = data.peekInt<uint8_t>(2);
+  is_two_way_ = (flag & TwoWayMask) == TwoWayMask ? true : false;
+  type_ = static_cast<SerializationType>(flag & SerializationTypeMask);
+  if (!isValidSerializationType(type_)) {
+    throw EnvoyException(
+        fmt::format("invalid dubbo message serialization type {}",
+                    static_cast<std::underlying_type<SerializationType>::type>(type_)));
+  }
+}
+
+void ResponseMessageImpl::fromBuffer(Buffer::Instance& buffer) {
+  ASSERT(buffer.length() >= DubboProtocolImpl::MessageSize);
+  status_ = static_cast<ResponseStatus>(buffer.peekInt<uint8_t>(3));
+  if (!isValidResponseStatus(status_)) {
+    throw EnvoyException(
+        fmt::format("invalid dubbo message response status {}",
+                    static_cast<std::underlying_type<ResponseStatus>::type>(status_)));
+  }
+}
+
+bool DubboProtocolImpl::decode(Buffer::Instance& buffer, Protocol::Context* context) {
+  if (buffer.length() < DubboProtocolImpl::MessageSize) {
+    return false;
+  }
+
+  uint16_t magic_number = buffer.peekBEInt<uint16_t>();
+  if (magic_number != MagicNumber) {
+    throw EnvoyException(fmt::format("invalid dubbo message magic number {}", magic_number));
+  }
+
+  uint8_t flag = buffer.peekInt<uint8_t>(2);
+  MessageType type = static_cast<MessageType>((flag & MessageTypeMask) >> 7);
```

Comment:
nit: I would prefer `(flag & MessageTypeMask) == MessageTypeMask ? Request : Response`, up to you